server/routes/api/settings.ts

```diff
@@ -129,41 +76,57 @@ settingsRoutes.route("/qr_enabled")
 
 settingsRoutes.route("/branch_roles")
 	.get(isAdmin, async (request, response) => {
-		let branchNames = (await validateSchema(config.questions, "./config/questions.schema.json")).map(branch => branch.name);
-		let applicationBranches = await getSetting<string[]>("applicationBranches");
-		let confirmationBranches = await getSetting<string[]>("confirmationBranches");
 		response.json({
-			"noop": branchNames.filter(branchName => applicationBranches.indexOf(branchName) === -1 && confirmationBranches.indexOf(branchName) === -1),
-			"applicationBranches": applicationBranches,
-			"confirmationBranches": confirmationBranches,
-			"applicationToConfirmationMap": await getSetting<ApplicationToConfirmationMap>("applicationToConfirmation")
+			"noop": (await Branches.BranchConfig.loadAllBranches("Noop")).map(branch => branch.name),
+			"applicationBranches": (await Branches.BranchConfig.loadAllBranches("Application")).map(branch => branch.name),
+			"confirmationBranches": (await Branches.BranchConfig.loadAllBranches("Confirmation")).map(branch => branch.name)
 		});
 	})
 	.put(isAdmin, uploadHandler.any(), async (request, response) => {
-		// First extract the application to confirmation map
-		let applicationToConfirmationMap: ApplicationToConfirmationMap = JSON.parse(request.body.applicationToConfirmationMap);
-		delete request.body.applicationToConfirmationMap;
-
-		let applicationBranches = [];
-		let confirmationBranches = [];
 		if ((new Set(Object.keys(request.body))).size !== Object.keys(request.body).length) {
 			response.status(400).json({
 				"error": "Each branch can only be used once"
 			});
 			return;
 		}
-		for (let branchName of Object.keys(request.body)) {
-			if (request.body[branchName] === "application") {
-				applicationBranches.push(branchName);
-			}
-			if (request.body[branchName] === "confirmation") {
-				confirmationBranches.push(branchName);
-			}
-		}
+
 		try {
-			await updateSetting<string[]>("applicationBranches", applicationBranches);
-			await updateSetting<string[]>("confirmationBranches", confirmationBranches);
-			await updateSetting<ApplicationToConfirmationMap>("applicationToConfirmation", applicationToConfirmationMap);
+			// TODO use promises/async
+			for (let branchName of Object.keys(request.body)) {
+				let branchData = JSON.parse(request.body[branchName]);
+
+				let branch = await Branches.BranchConfig.loadBranchFromDB(branchName); // TODO type checker determines this to be always a NoopBranch - ensure that branch.type will be the real type everytime
```

Comment:
It's saying that this is always a NoopBranch because that is the superclass of all other branches @bunsenmcdubbs

Reply:
I assume that `branch.type` is correct even if the type system determines that this is a `NoopBranch`. is that a safe assumption? aka does line 100 work as expected? @petschekr